### PR TITLE
feat: get last bitcoin fees from sweep tables (#809)

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

Closes: #809

## Changes

Removes the `BitcoinInteract::get_last_fees()` (described in #541, replaced by #809) and replaces it with logic to retrieve the `Fees` from the latest sweep transaction package.

## Testing Information

RBF tests updated to use the `sweep_transactions` method.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/transaction_coordinator.rs [hunk 1]
```diff
@@ -351,12 +352,6 @@ where
         let transaction_package = pending_requests.construct_transactions()?;
 
         for mut transaction in transaction_package {
-            self.send_message(
```
Comment (cylewitruk) [inline signer/src/transaction_coordinator.rs:354]: This is moved now because we need to store the size of the transaction after signing, which happens in `sign_and_broadcast()`.

File: signer/src/transaction_coordinator.rs [hunk 2]
```diff
@@ -678,12 +673,20 @@ where
...
-
+        // Broadcast the transaction to the Bitcoin network.
         self.context
             .get_bitcoin_client()
             .broadcast_transaction(&transaction.tx)
             .await?;
 
+        // Publish the transaction to the P2P network so that peers get advance
+        // knowledge of the sweep.
+        self.send_message(
```
Comment (cylewitruk) [inline signer/src/transaction_coordinator.rs:684]: Tbh I'm not sure which is better, publishing this before or after the bitcoin broadcast.. :thinking:

File: signer/migrations/0003__create_tables.sql [hunk 3]
```diff
@@ -196,6 +196,8 @@ CREATE TABLE sbtc_signer.sweep_transactions (
     amount BIGINT NOT NULL,
     -- The fee paid for the transaction.
     fee BIGINT NOT NULL,
+    -- The _vsize_ of the transaction in bytes.
```
Comment (djordon) [inline signer/migrations/0003__create_tables.sql:199]: I would rephrase this to say something like 'The bitcoin "virtual size" of the transaction'. I don't think that saying that it is in bytes is incorrect.

File: signer/src/message.rs [hunk 4]
```diff
@@ -164,6 +166,10 @@ pub struct SweepTransactionInfo {
 impl SweepTransactionInfo {
     /// Creates a [`SweepTransactionInfo`] from an [`UnsignedTransaction`] and a
     /// Bitcoin block hash.
+    ///
+    /// **Note that it is important that this function is called after the
+    /// transaction has been signed**, as otherwise the `vbytes` size will not
+    /// be correct.
```
Comment (djordon) [inline signer/src/message.rs:172]: When creating an `UnsignedTransaction` we compute what the `vsize` would be if it were signed, but then throw that value away since we didn't need it at the time. We could keep it around if we want to.

File: signer/src/message.rs [hunk 5]
```diff
@@ -208,6 +214,8 @@ impl SweepTransactionInfo {
                 .signers_script_pubkey(),
             amount: unsigned.output_amounts(),
             fee: unsigned.tx_fee,
+            // the vsize should never be larger than u32::MAX
```
Comment (djordon) [inline signer/src/message.rs:217]: Oh because the max block vsize is currently `u32::MAX` bytes, right?

File: signer/tests/integration/rbf.rs [hunk 6]
```diff
@@ -188,7 +192,7 @@ pub fn transaction_with_rbf(
         })
         .collect();
 
-    faucet.generate_blocks(1);
+    let block_hash = faucet.generate_blocks(1);
```
Comment (djordon) [inline signer/tests/integration/rbf.rs:195]: nit: I would add a `.pop().unwrap()` here, since it makes reading line 260 a little easier.

File: signer/migrations/0003__create_tables.sql [hunk 7]
```diff
@@ -196,6 +196,8 @@ CREATE TABLE sbtc_signer.sweep_transactions (
...
+    vsize INTEGER NOT NULL,
```
Comment (djordon) [inline signer/migrations/0003__create_tables.sql:200]: We should make this a `BIGINT` since the max is `u32::MAX`. Hmmm, it feels like a waste of space, but it is more correct.

File: signer/src/storage/postgres.rs [hunk 8]
```diff
@@ -1749,6 +1750,60 @@ impl super::DbRead for PgStore {
...
+                    sweep_txs last 
+                        ON tx.signer_prevout_txid = last.txid
+            )
+            SELECT
+                txid
+              , number
+            FROM
+                sweep_txs
+            ORDER BY
+                number ASC;
```
Comment (djordon) [inline signer/src/storage/postgres.rs:1788]: Doesn't this find all transactions after the `prevout_txid`? I mean, including the ones issued as part of subsequent transaction packages?

Comment (djordon) [reply signer/src/storage/postgres.rs:1788]: Also, this returns the transaction that includes the given txid. This isn't bad per se, but when computing the last fee we'll include a transaction that has been confirmed, which will give us a bad last fee. And if we have RBF'ed a transaction package more than once, this will include all packages not just the one that we care about, which is the last one.

**Review by djordon — COMMENTED** [review]
This looks okay, but I don't think the query does what we want (or we need to change our usage).

File: signer/src/storage/postgres.rs (hunk 8 above)
Comment (cylewitruk) [reply signer/src/storage/postgres.rs:1788]: Yeah you're right, there's some funk here. I'll add a couple negative tests to ensure those scenarios can't happen and update the query.

File: signer/src/message.rs (hunk 5 above)
Comment (cylewitruk) [reply signer/src/message.rs:217]: That and that it would be completely insane if we had a transaction of this size; max bitcoin block size is ~4mb.

File: signer/src/storage/postgres.rs (hunk 8 above)
Comment (cylewitruk) [reply signer/src/storage/postgres.rs]: I updated the query which I think solves the above issues (and current tests pass), but I still need to write the negative tests

File: signer/migrations/0003__create_tables.sql (hunk 3 above)
Comment (cylewitruk) [reply signer/migrations/0003__create_tables.sql:199]: Yeah you're right, since witness data is 1/4 weight

File: signer/src/message.rs (hunk 4 above)
Comment (cylewitruk) [reply signer/src/message.rs:172]: Ah good to know, I'll make that change

Comment (cylewitruk) [reply signer/src/message.rs]: Updated in [a399760](https://github.com/stacks-network/sbtc/pull/810/commits/a39976075a79f2c064b214d368485b787a1139e0)

File: signer/src/message.rs (hunk 5 above)
Comment (cylewitruk) [reply signer/src/message.rs]: This is changed in [a399760](https://github.com/stacks-network/sbtc/pull/810/commits/a39976075a79f2c064b214d368485b787a1139e0) from the above comment

File: signer/tests/integration/rbf.rs (hunk 6 above)
Comment (cylewitruk) [reply signer/tests/integration/rbf.rs]: Updated in [9b38cb1](https://github.com/stacks-network/sbtc/pull/810/commits/9b38cb1d33633cbc3655b95e53caa443d6e9ed16)

File: signer/migrations/0003__create_tables.sql (hunk 3 above)
Comment (cylewitruk) [reply signer/migrations/0003__create_tables.sql]: [77d4884](https://github.com/stacks-network/sbtc/pull/810/commits/77d4884094bc91a71e5c7ee44a96cc22b72b895e)

File: signer/migrations/0003__create_tables.sql (hunk 7 above)
Comment (djordon) [reply signer/migrations/0003__create_tables.sql:200]: Hmm, yeah. And we should probably cap it ourselves anyway.

File: signer/src/storage/mod.rs [hunk 9]
```diff
@@ -271,6 +271,13 @@ pub trait DbRead {
         chain_tip: &model::BitcoinBlockHash,
         context_window: u16,
     ) -> impl Future<Output = Result<Option<model::SweepTransaction>, Error>> + Send;
+
+    /// Get the sweep transaction package for the given previous transaction id
+    /// (i.e. the previous output that was spent from the then-UTXO).
```
Comment (djordon) [inline signer/src/storage/mod.rs:276]: I would add a little here and maybe rephrase it. I would say something like "This function takes the transaction ID of the signers' unspent output on the canonical Bitcoin blockchain and returns the chain of mempool transactions that spend that output." And maybe add something like this "The provided txid refers to an on-chain unspent output created by the signers, which may be spent by a transaction currently in the mempool."

Man that was wordy.

File: signer/src/transaction_coordinator.rs [hunk 10]
```diff
@@ -950,13 +953,28 @@ where
...
+            .get_storage()
+            .get_sweep_transaction_package(&utxo.outpoint.txid.into())
+            .await?;
+
+        // If the last sweep package is empty, then we don't have any fees to
+        // report, so we return None. Otherwise, we return the calculated fees
+        // from the last sweep package.
+        let last_fees = last_sweep_package
+            .is_empty()
+            .then(|| last_sweep_package.get_fees());
```
Comment (djordon) [inline signer/src/transaction_coordinator.rs:977]: I think this is backwards. If `last_sweep_package.is_empty()` then `last_fees` should be `None`

**Review by djordon — COMMENTED** [review]
I think there is only one issue with the `last_fee` part in the coordinator. Looks good otherwise.

File: signer/src/transaction_coordinator.rs (hunk 10 above)
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs]: I wasn't done! :stuck_out_tongue_closed_eyes: But yeah you're right, refactoring gone bad. But now there's tests that catch this and verify the whole `get_btc_state()` function both when sweeps exist and when they don't.
[6cf7d78](https://github.com/stacks-network/sbtc/pull/810/commits/6cf7d789d6db074533b7484aca5902d9eb310a99)

File: signer/src/storage/mod.rs (hunk 9 above)
Comment (cylewitruk) [reply signer/src/storage/mod.rs]: Updated the comments a bit in [0d4aed9](https://github.com/stacks-network/sbtc/pull/810/commits/0d4aed96c6ec3706a382ee98797298178da90c9f).

While what you describe is the _intended use_, it's not really true for the provided functionality of the method. It purposely places the burden of making sure you're providing the correct `prevout_txid` on the caller and thus itself provides no guarantees of canonicality or that the `prevout_txid` is indeed confirmed.

Comment (cylewitruk) [reply signer/src/storage/mod.rs]: Actually renamed the method in [413ae4c](https://github.com/stacks-network/sbtc/pull/810/commits/413ae4ce3d7195a326e8a9695866d605d5ec9963) to better reflect what it does

File: signer/src/transaction_coordinator.rs (hunk 10 above)
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Oh whoops, I didn't check to see if you were done yet.

**Review by djordon — APPROVED** [review]
Looks good ✅ 🚢. We should check if the integration tests in the `transaction_coordinator` run in CI.

**Comment by djordon** [comment]
I decided to go and check myself and I didn't find those tests in the CI output logs. Also, we should have CI explicitly output which tests were ignored.

File: signer/src/transaction_coordinator.rs [hunk 11]
```diff
@@ -1380,4 +1413,295 @@ mod tests {
     async fn should_get_signer_utxo_donations() {
         test_environment().assert_get_signer_utxo_donations().await;
     }
+
+    /// This test asserts that the `get_btc_state` function returns the correct
+    /// `SignerBtcState` when there are no sweep transactions available, i.e.
+    /// the `last_fees` field should be `None`.
+    #[cfg_attr(not(feature = "integration-tests"), ignore)]
+    #[test(tokio::test)]
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1421]: It's from `test_log::test` -- lets you run the test with `RUST_LOG=xx` and get actual logs output. You'll find it on a number of tests which I've needed to debug :sweat_smile:

File: signer/src/transaction_coordinator.rs [hunk 12]
```diff
@@ -1380,4 +1413,295 @@ mod tests {
     async fn should_get_signer_utxo_donations() {
         test_environment().assert_get_signer_utxo_donations().await;
     }
+
+    /// This test asserts that the `get_btc_state` function returns the correct
+    /// `SignerBtcState` when there are no sweep transactions available, i.e.
+    /// the `last_fees` field should be `None`.
+    #[cfg_attr(not(feature = "integration-tests"), ignore)]
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1420]: @djordon hmm, I'm not sure of the best approach. Here we're testing `get_btc_state()` which is private, so it needs to be here, but otoh it also needs postgres, so it needs to be run as an integration-test. Should I just make it public and move the test, or do you have a better suggestion?

File: signer/src/storage/in_memory.rs [hunk 13]
```diff
@@ -810,6 +810,31 @@ impl super::DbRead for SharedStore {
 
         Ok(package)
     }
+
+    async fn get_latest_unconfirmed_sweep_transactions(
+        &self,
+        prevout_txid: &model::BitcoinTxId,
+    ) -> Result<Vec<model::SweepTransaction>, Error> {
```
Comment (matteojug) [inline signer/src/storage/in_memory.rs:817]: nit: wrt the pg one (and the trait docstring), this doesn't filter out confirmed ones IIUC. I guess it's fine, but maybe we should add a comment about it and revisit later?

File: signer/src/storage/model.rs [hunk 14]
```diff
@@ -84,6 +91,17 @@ impl From<&crate::message::SweepTransactionInfo> for SweepTransaction {
     }
 }
 
+impl utxo::GetFees for Vec<SweepTransaction> {
+    /// Return the total fee of all the transactions in the vector.
+    fn get_fees(&self) -> Fees {
+        let total = self.iter().map(|tx| tx.fee).sum();
+        let total_size: u32 = self.iter().map(|tx| tx.vsize).sum();
```
Comment (matteojug) [inline signer/src/storage/model.rs:98]: qq: do we risk to overflow when summing `vsize`s?

File: signer/src/storage/model.rs [hunk 15]
```diff
@@ -84,6 +91,17 @@ impl From<&crate::message::SweepTransactionInfo> for SweepTransaction {
...
+        let rate: f64 = total as f64 / total_size as f64;
```
Comment (matteojug) [inline signer/src/storage/model.rs:99]: should we add a `total_size == 0` check here? I guess it shouldn't happen, merely as a safety guard

File: signer/src/storage/postgres.rs [hunk 16]
```diff
@@ -1749,6 +1750,88 @@ impl super::DbRead for PgStore {
...
+        // chained from the first one. As a precaution, this query also filters
+        // out any sweep transactions that have already been confirmed on the
+        // Bitcoin blockchain.
+        let infos: Vec<(model::BitcoinTxId, i32)> = sqlx::query_as(
+            "
+            WITH RECURSIVE sweep_txs AS (
+                SELECT
+                    txid
+                  , signer_prevout_txid
+                  , 1 AS number
```
Comment (matteojug) [inline signer/src/storage/postgres.rs:1792]: nit: do we need the number?

File: signer/src/storage/postgres.rs [hunk 17]
```diff
@@ -1749,6 +1750,88 @@ impl super::DbRead for PgStore {
...
+                FROM sweep_transactions tx
+                INNER JOIN sweep_txs last 
+                    ON tx.signer_prevout_txid = last.txid
+            )
+            SELECT
+                sweep_txs.txid
+              , number
+            FROM sweep_txs
+            LEFT JOIN bitcoin_transactions AS btc_tx 
+                ON btc_tx.txid = sweep_txs.txid
```
Comment (matteojug) [inline signer/src/storage/postgres.rs:1812]: What happens in case of forks here? eg a sweep tx is confirmed on a block, but then we get two block on another fork (making this second one the canonical one): the sweep tx may still be in the mempool for this second (canonical) fork, but here we would exclude it.

**Comment by djordon** [comment]
> @djordon hmm, I'm not sure of the best approach. Here we're testing `get_btc_state()` which is private, so it needs to be here, but otoh it also needs postgres, so it needs to be run as an integration-test. Should I just make it public and move the test, or do you have a better suggestion?

Yeah I'd make it public.

File: signer/src/storage/model.rs (hunk 14 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs:98]: In theory yes, but then we're up to 4MB worth of transaction data (well, roughly -- witness data is discounted to 1/4). So in practical terms, this should never happen.

But it's a good point, I'll refactor this a little to return a result, I guess it could potentially happen with `Fake`'d stuff.

File: signer/src/storage/postgres.rs (hunk 17 above)
Comment (djordon) [reply signer/src/storage/postgres.rs:1812]: Ahhh, good point... we need to join on the canonical bitcoin blockchain... nuts.

File: signer/src/storage/model.rs (hunk 15 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs:99]: Right, I don't think it's even possible to create a zero-sized `bitcoin::Transaction` -- but I'll toss that in while I'm doing the above refactor just so it's 100% sound.

File: signer/src/storage/in_memory.rs (hunk 13 above)
Comment (cylewitruk) [reply signer/src/storage/in_memory.rs:817]: The
```sql
LEFT JOIN bitcoin_transactions AS btc_tx 
    ON btc_tx.txid = sweep_txs.txid
WHERE btc_tx.txid IS NULL
```
should take care of that wrt the sweep transactions themselves. If you're thinking about the `prevout_txid` which you pass in, it's expected that you as the caller have done the work to ensure that that transaction you're referring to is correct.

File: signer/src/network/in_memory2.rs [hunk 18]
```diff
@@ -83,10 +87,22 @@ impl SignerNetwork {
...
+    /// Create a new in-memory signer network with a single signer instance.
+    /// You can use this if you do not need to simulate multiple signers.
+    pub fn single() -> Self {
+        let (wan_tx, _) = tokio::sync::broadcast::channel(DEFAULT_WAN_CAPACITY);
+        let (signer_tx, _) = tokio::sync::broadcast::channel(DEFAULT_SIGNER_CAPACITY);
+        Self {
+            wan_tx,
+            signer_tx,
+            sent: Arc::new(RwLock::new(VecDeque::new())),
+        }
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/network/in_memory2.rs]: Haha, yes that works too :angel: [f9c4ead](https://github.com/stacks-network/sbtc/pull/810/commits/f9c4ead98c5c5d68bec6ccae0ce37b31b0c45354)

File: signer/src/storage/in_memory.rs (hunk 13 above)
Comment (matteojug) [reply signer/src/storage/in_memory.rs:817]: Uhm I'm not sure I'm following: here I was referring to the in-memory impl not filtering out the confirmed sweep tx, while the pg does it (correctly, according to the docstring). Am I missing something?

File: signer/src/storage/postgres.rs (hunk 16 above)
Comment (matteojug) [reply signer/src/storage/postgres.rs:1792]: Ah I missed that, now it makes sense 😅

File: signer/src/storage/postgres.rs (hunk 17 above)
Comment (cylewitruk) [reply signer/src/storage/postgres.rs:1812]: Welp... I guess sooner or later someone was going to have to write a canonical function for transactions... :tada:

File: signer/src/storage/in_memory.rs (hunk 13 above)
Comment (cylewitruk) [reply signer/src/storage/in_memory.rs:817]: Ohh sorry I thought you meant with regards to the PG one. Right, i'll have to check the in-mem one or just mark it `unimplemented!()`, I think I forgot to circle back to that impl after the latest updates.

File: signer/src/storage/model.rs (hunk 15 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs:99]: That too, so good catch

File: signer/src/transaction_coordinator.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs:1421]: 👍🏽 good to know.

File: signer/src/storage/model.rs (hunk 14 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs]: Refactored in [8018721](https://github.com/stacks-network/sbtc/pull/810/commits/8018721e7104fdbe5b368af70835ffbc23f437c5)

File: signer/src/storage/model.rs (hunk 15 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs:109]: Added in [8018721](https://github.com/stacks-network/sbtc/pull/810/commits/8018721e7104fdbe5b368af70835ffbc23f437c5)

File: signer/src/storage/model.rs [hunk 19]
```diff
@@ -93,12 +93,22 @@ impl From<&crate::message::SweepTransactionInfo> for SweepTransaction {
...
     /// Return the total fee of all the transactions in the vector.
-    fn get_fees(&self) -> Fees {
-        let total = self.iter().map(|tx| tx.fee).sum();
-        let total_size: u32 = self.iter().map(|tx| tx.vsize).sum();
+    fn get_fees(&self) -> Result<Fees, Error> {
+        let total: u64 = self
+            .iter()
+            .map(|tx| tx.fee)
+            .try_fold(0u64, |acc, fee| acc.checked_add(fee))
+            .ok_or(Error::ArithmeticOverflow)?;
```
Comment (djordon) [inline signer/src/storage/model.rs:101]: Hmmm, these fees cannot add up to anything more than `u64::MAX`, even if we added up all bitcoins (measured in sats).

File: signer/src/storage/model.rs [hunk 20]
```diff
@@ -93,12 +93,22 @@ impl From<&crate::message::SweepTransactionInfo> for SweepTransaction {
...
+
+        let total_size: u64 = self.iter().map(|tx| tx.vsize as u64).sum();
+
+        if total_size == 0 {
+            return Err(Error::DivideByZero);
+        }
```
Comment (djordon) [inline signer/src/storage/model.rs:111]: This can only happen if the vector is empty, so we could just return zero.

File: signer/src/storage/model.rs (hunk 19 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs:101]: Nope, but the code otherwise is _technically_ not panic-safe, which is what we've been striving for.

File: signer/src/storage/model.rs (hunk 20 above)
Comment (cylewitruk) [reply signer/src/storage/model.rs]: I don't think zero is what we want -- `compute_transaction_fee()` matches on an `Option<>` and `None` would not be the same as `Fees::ZERO`. 

But I changed this to return `None` ([891e46a](https://github.com/stacks-network/sbtc/pull/810/commits/891e46aa7ce8aaf05a0fb7e87912aca2a5e96b05)) if `total_size` is zero since I could describe that mode in the trait description and it matches its use. Yeah it should only happen if the vec is empty in prod code, but it's still not panic-safe by simply checking `self.len()` (you could create items all of which have `vsize == 0`).

File: signer/src/storage/postgres.rs [hunk 21]
```diff
@@ -1701,6 +1702,99 @@ impl super::DbRead for PgStore {
 
         self.get_sweep_transaction_by_txid(&txid).await
     }
+
+    async fn get_latest_unconfirmed_sweep_transactions(
+        &self,
+        chain_tip: &model::BitcoinBlockHash,
+        context_window: u16,
+        prevout_txid: &model::BitcoinTxId,
+    ) -> Result<Vec<model::SweepTransaction>, Error> {
```
Comment (matteojug) [inline signer/src/storage/postgres.rs:1711]: nit: this fn may return a suffix of the chain of sweep tx starting from `prevout_txid`, if some prefix of the chain is confirmed. The result is that the first sweep tx of the vec may not be spending `prevout_txid`. I guess it's fine with its current usage, but I wanted to point it out in case we plan to rely on that assumption (feel free to resolve this convo).

**Review by matteojug — APPROVED** [review]
LGTM, but I think you should revert the changes in generated code (`make lint` should do the trick).

**Comment by cylewitruk** [comment]
@matteojug yeah, crap, those generated files snuck in on me. `make lint` did indeed seem to resolve it :+1:

File: signer/src/storage/model.rs (hunk 20 above)
Comment (djordon) [reply signer/src/storage/model.rs:120]: Great points!

File: signer/tests/integration/postgres.rs [hunk 22]
```diff
@@ -2311,6 +2314,246 @@ async fn can_store_and_get_latest_sweep_transaction() {
...
+    assert_eq!(sweep_transactions.len() - 1, sweeps.len());
+    assert_eq!(&sweep_transactions[1..], &sweeps[..]);
+
+    // =========================================================================
+    // Step 5 - create a fork from block 1, competing with block "2a"
+    // -------------------------------------------------------------------------
+    // We create two new Bitcoin blocks, chained off of the first Bitcoin block
+    // and thus competing with block 2a. By creating two new blocks, this chain
+    // now becomes the new canonical (longest) chain.
+    // =========================================================================
```
Comment (djordon) [inline signer/tests/integration/postgres.rs:2518]: I like this style of comment a lot. 👍🏽

**Review by djordon — APPROVED** [review]
